options/options.go

```diff
@@ -795,5 +803,245 @@ type EngineOptions struct {
+				// try again
+			case <-ctx.Done():
+				return errors.New(ctx.Err())
+			}
+
+			currentAttempt++
+
+			continue
+		}
+
+		return err
+	}
+}
+
+func (opts *TerragruntOptions) handleIgnoreSignals(signals map[string]interface{}) error {
+	workingDir := opts.WorkingDir
+	signalsFile := filepath.Join(workingDir, DefaultSignalsFile)
+	signalsJSON, err := json.MarshalIndent(signals, "", "  ")
+
+	if err != nil {
+		return err
+	}
+
+	const ownerPerms = 0644
+	if err := os.WriteFile(signalsFile, signalsJSON, ownerPerms); err != nil {
+		return fmt.Errorf("failed to write signals file %s: %w", signalsFile, err)
+	}
+
+	opts.Logger.Warnf("Written error signals to %s", signalsFile)
+
+	return nil
+}
+
+// ErrorAction represents the action to take when an error occurs
+type ErrorAction struct {
+	ShouldIgnore   bool
+	ShouldRetry    bool
+	IgnoreMessage  string
+	IgnoreSignals  map[string]interface{}
+	RetryMessage   string
+	RetryAttempts  int
+	RetrySleepSecs int
+}
+
+// ProcessError evaluates an error against the configuration and returns the appropriate action
+func (c *ErrorsConfig) ProcessError(err error, currentAttempt int) (*ErrorAction, error) {
+	if err == nil {
+		return nil, nil
+	}
+
+	errStr := err.Error()
+	action := &ErrorAction{}
+
+	// First check ignore rules
+	for _, ignoreBlock := range c.Ignore {
+		isIgnorable, err := matchesAnyPattern(errStr, ignoreBlock.IgnorableErrors)
+		if err != nil {
+			return nil, errors.New(fmt.Sprintf("error processing ignore patterns: %v", err))
+		}
+
+		if isIgnorable {
+			action.ShouldIgnore = true
+			action.IgnoreMessage = ignoreBlock.Message
+			action.IgnoreSignals = make(map[string]interface{})
+
+			// Convert cty.Value map to regular map
+			for k, v := range ignoreBlock.Signals {
+				action.IgnoreSignals[k] = v
+			}
+
+			return action, nil
+		}
+	}
+
+	// Then check retry rules
+	for _, retryBlock := range c.Retry {
+		isRetryable, err := matchesAnyPattern(errStr, retryBlock.RetryableErrors)
+		if err != nil {
+			return nil, errors.New(fmt.Sprintf("error processing retry patterns: %v", err))
+		}
+
+		if isRetryable {
+			if currentAttempt >= retryBlock.MaxAttempts {
+				return nil, errors.New(fmt.Sprintf("max retry attempts (%d) reached for error: %v",
+					retryBlock.MaxAttempts, err))
+			}
+
+			action.RetryMessage = retryBlock.Name
+			action.ShouldRetry = true
+			action.RetryAttempts = retryBlock.MaxAttempts
+			action.RetrySleepSecs = retryBlock.SleepIntervalSec
+
+			return action, nil
+		}
+	}
+
+	return nil, err
+}
+
+// matchesAnyPattern checks if the input string matches any of the provided patterns
+func matchesAnyPattern(input string, patterns []string) (bool, error) {
+	for _, pattern := range patterns {
+		// Handle negative patterns (patterns starting with !)
+		isNegative := false
+		if len(pattern) > 0 && pattern[0] == '!' {
+			isNegative = true
+			pattern = pattern[1:]
+		}
+
+		matched, err := regexp.MatchString(pattern, input)
```

Comment:
In addition to the comment above, note that it's also more efficient to have this compiled earlier so that we don't recompile all the patterns for every error.